stream.go

```diff
@@ -155,6 +160,11 @@ func (cs *clientStream) Trailer() metadata.MD {
 }
 
 func (cs *clientStream) SendMsg(m interface{}) (err error) {
+	cs.mu.Lock()
```

Comment:
I agree this makes the code look better. But I still think it should NOT
impact the performance of the fast path. It is understood it is per-stream
lock and there is much contention on it. But it still adds at least a few
hundreds of nanoseconds (according to the experience with C++ code, it is
probably longer in Go) even though there is no contention. It is likely
invisible in the current benchmark (unless some deliberate benchmark is
designed) but it could show up when we introduce some other high-speed
transport (e.g., shared memory, RDMA) in the future. I hope we try our best
to hold "the fast path should not be impacted" principle across the entire
dev process.

On Tue, Jun 30, 2015 at 6:36 PM, David Symonds <notifications@github.com>
wrote:

> In stream.go
> <https://github.com/grpc/grpc-go/pull/235#discussion_r33641754>:
>
> > @@ -155,6 +160,11 @@ func (cs *clientStream) Trailer() metadata.MD {
> >  }
> >
> >  func (cs *clientStream) SendMsg(m interface{}) (err error) {
> > +	cs.mu.Lock()
>
> I thought this approach was better since it limited the number of places
> touching the global EnableTracing var.
>
> Keep in mind that this is a per-stream lock, not a global lock, so it only
> is contended when the stream itself is busy. That's not often the case
> (since streams usually require a specific ordering), so the mutex overhead
> will normally be unnoticed. I can run the benchmarks if you'd like me to
> verify this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/235/files#r33641754>.
>